src/frequenz/sdk/timeseries/_formula_engine/_formula_engine.py

```diff
@@ -473,76 +495,85 @@ def push_average(self, metrics: List[Tuple[str, Receiver[Sample], bool]]) -> Non
             fetchers.append(fetcher)
         self._steps.append(Averager(fetchers))
 
-    def build(self) -> FormulaEngine:
-        """Finalize and build the formula engine.
+    @property
+    def name(self) -> str:
+        """Return the name of the formula being built.
 
         Returns:
-            A `FormulaEngine` instance.
+            The name of the formula being built.
+        """
+        return self._name
+
+    async def subscribe(self) -> None:
+        """Subscribe to metrics if needed.
+
+        This is a no-op for the `FormulaBuilder` class, but is used by the
+        `ResampledFormulaBuilder` class.
+        """
+
+    def finalize(self) -> tuple[list[FormulaStep], dict[str, MetricFetcher]]:
+        """Finalize and return the steps and fetchers for the formula.
+
+        Returns:
+            A tuple of the steps and fetchers for the formula.
         """
         while self._build_stack:
             self._steps.append(self._build_stack.pop())
 
-        return FormulaEngine(self._name, self._steps, self._metric_fetchers)
+        return self._steps, self._metric_fetchers
 
+    def build(self) -> FormulaEngine:
+        """Create a formula engine with the steps and fetchers that have been pushed.
 
-_GenericSample = TypeVar("_GenericSample", Sample, Sample3Phase)
-_GenericEngine = TypeVar("_GenericEngine", FormulaEngine, FormulaEngine3Phase)
-_GenericFormulaChannel = TypeVar(
-    "_GenericFormulaChannel", "FormulaChannel", "FormulaChannel3Phase"
-)
-_GenericFormulaReceiver = TypeVar(
-    "_GenericFormulaReceiver", "FormulaReceiver", "FormulaReceiver3Phase"
-)
-_GenericHOFormulaBuilder = TypeVar(
-    "_GenericHOFormulaBuilder",
-    "HigherOrderFormulaBuilder",
-    "HigherOrderFormulaBuilder3Phase",
-)
+        Returns:
+            A `FormulaEngine` instance.
+        """
+        self.finalize()
+        return FormulaEngine(self)
 
 
-class _BaseHOFormulaBuilder(
-    ABC, Generic[_GenericFormulaReceiver, _GenericSample, _GenericEngine]
-):
+class _BaseHOFormulaBuilder(ABC):
     """Provides a way to build formulas from the outputs of other formulas."""
 
-    def __init__(self, recv: _GenericFormulaReceiver) -> None:
+    def __init__(self, engine: FormulaEngine | FormulaEngine3Phase) -> None:
         """Create a `GenericHigherOrderFormulaBuilder` instance.
 
         Args:
-            recv: A first input stream to create a builder with, so that python
+            engine: A first input stream to create a builder with, so that python
                 operators `+, -, *, /` can be used directly on newly created instances.
         """
-        self._steps: deque[tuple[TokenType, _GenericFormulaReceiver | str]] = deque()
-        self._steps.append((TokenType.COMPONENT_METRIC, recv.clone()))
-        recv._deactivate()  # pylint: disable=protected-access
-        self._engine = None
+        self._engine: FormulaEngine | FormulaEngine3Phase
```

Comment:
is it really needed to keep the state for this attribute?

Reply:
true,  it is not necessary.  removed.